core/app/models/spree/order/checkout.rb

```diff
@@ -70,12 +70,18 @@ def self.define_state_machine!
               end
 
               if states[:payment]
+                event :payment_failed do
+                  transition to: :payment, from: :confirm
+                end
+
                 before_transition to: :complete do |order|
                   if order.payment_required? && order.payments.valid.empty?
                     order.errors.add(:base, Spree.t(:no_payment_found))
                     false
                   elsif order.payment_required?
-                    order.process_payments!
+                    order.process_payments!.tap do |success|
+                      order.handle_failed_payments unless success
+                    end
```

Comment:
I don't think the business logic should be placed here. The state machine is anyhow overly complicated.

There was an attempt somewhere, I think by @athal7 , to create Ruby class for processing payments. That should be the correct place for that kind of logic.